### PR TITLE
Improves seq2seq example

### PR DESCRIPTION
**Pull request by marcvanzee**
Makes the following improvements:

- Factors out model and input pipeline code into separate files. Before it was a single file, but it was actually too much code so it made it hard to navigate around
- Splits rngs inside jitted function
- No longer use a global CharacterTable, but pass around a local one instead.
- Fixes tests with new setup

Further follow-up improvements:

- Update Colab
- Put relevant stuff of CharacterTable inside a train_state so we only have to pass around that (and it can also be passed to jitted functions)
- Add type annotations everywhere

### CODE REVIEW

File: examples/seq2seq/input_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Input pipeline for seq2seq addition example."""
+
+import numpy as np
```
Comment (andsteing) [inline examples/seq2seq/input_pipeline.py:17]: nit: first standard library, then other packages, i.e.

```
import random


import jax
import jax.numpy as jnp
import numpy as np
```

File: examples/seq2seq/input_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import random
+
+import jax
+import jax.numpy as jnp
+
+
+class CharacterTable(object):
```
Comment (andsteing) [inline examples/seq2seq/input_pipeline.py:24]: "new style classes" is a Python 2.x thing ...

File: examples/seq2seq/input_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  def encoder_input_shape(self):
+    max_input_len = self._max_len_query_digit * 2 + 2  # includes EOS
+    return (1, max_input_len, self.vocab_size)
+
+  @property
+  def decoder_input_shape(self):
+    return (1, self.max_output_len, self.vocab_size)
+
+  def encode(self, inputs):
+    """Encode from string to list of integers."""
```
Comment (andsteing) [inline examples/seq2seq/input_pipeline.py:69]: nit: function pydoc is "descriptive style" i.e. `Encodes` and not `Encode` (here and elsewhere)

File: examples/seq2seq/input_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    return np.array([encode_str(inp) for inp in batch_inputs])
+
+  def decode_onehot(self, batch_inputs):
+    """Decode a batch of one-hot encoding to strings."""
+    decode_inputs = lambda inputs: self.decode(inputs.argmax(axis=-1))
+    return np.array(list(map(decode_inputs, batch_inputs)))
+
+  def generate_examples(self, num_examples):
+    """Returns @num_examples examples."""
```
Comment (andsteing) [inline examples/seq2seq/input_pipeline.py:104]: nit: `Yields`

File: examples/seq2seq/models.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    carried_lstm_state = tuple(
+        select_carried_state(*s) for s in zip(new_lstm_state, lstm_state))
+    # Update `is_eos`.
+    is_eos = jnp.logical_or(is_eos, x[:, self.eos_id])
+    return (carried_lstm_state, is_eos), y
+
+
+  @staticmethod
+  def initialize_carry(batch_size, hidden_size):
+    # use dummy key since default state init fn is just zeros.
```
Comment (andsteing) [inline examples/seq2seq/models.py:59]: Would `None` work instead?

File: examples/seq2seq/models.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    return nn.LSTMCell.initialize_carry(
+        jax.random.PRNGKey(0), (batch_size,), hidden_size)
+
+
+class Encoder(nn.Module):
+  """LSTM encoder, returning state after EOS is input."""
```
Comment (andsteing) [inline examples/seq2seq/models.py:65]: Can't parse this sentence; should `is` be `in` ?

**Review by andsteing — APPROVED** [review]
Thanks for updating this example!

Added mostly small nits/suggestions.

File: examples/seq2seq/input_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply examples/seq2seq/input_pipeline.py:104]: Hmm no me neither, removed it.

File: examples/seq2seq/models.py (hunk 5 above)
Comment (marcvanzee) [reply examples/seq2seq/models.py:59]: No, because LSTMCell.initialize_carry splits the rng.

File: examples/seq2seq/models.py (hunk 6 above)
Comment (marcvanzee) [reply examples/seq2seq/models.py:65]: input is used as a verb, but I'll rephrase it to make it more clear.

File: examples/seq2seq/models.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import functools
+from typing import Any, Tuple
+
+from flax import linen as nn
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+PRNGKey = Any
+Shape = Tuple[int]
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply examples/seq2seq/models.py:29]: Hmm no actually none of these seem to be used. Removed them.
